x-pack/plugins/fleet/server/services/security/uninstall_token_service/index.ts

```diff
@@ -82,62 +156,140 @@ export class UninstallTokenService implements UninstallTokenServiceInterface {
+          sortField: 'created_at',
+          sortOrder: 'desc',
+        }
+      );
+    let decryptedTokenObjects: Array<SavedObjectsFindResult<UninstallTokenSOAttributes>> = [];
 
-  /**
-   * gets uninstall tokens for given policy ids
-   *
-   * @param policyIds agent policy ids
-   * @returns array of UninstallToken objects
-   */
-  public async getTokensForPolicyIds(policyIds: string[]): Promise<UninstallToken[]> {
-    return (await this.getTokensByIncludeFilter({ include: policyIds })).items;
+    for await (const result of tokensFinder.find()) {
+      decryptedTokenObjects = result.saved_objects;
+      break;
+    }
+    tokensFinder.close();
+
+    const decryptedTokens = decryptedTokenObjects.map(({ created_at: createdAt, attributes }) => {
+      this.assertPolicyId(attributes);
+      this.assertToken(attributes);
+      this.assertCreatedAt(createdAt);
+
+      return {
+        policy_id: attributes.policy_id,
+        token: attributes.token || attributes.token_plain,
+        created_at: createdAt,
+      };
+    });
+
+    return {
+      items: decryptedTokens,
+      total: decryptedTokenObjects.length,
+    };
   }
-  /**
-   * gets uninstall token for given policy id, paginated
-   *
-   * @param searchString a string for partial matching the policyId
-   * @param page
-   * @param perPage
-   * @param policyId agent policy id
-   * @returns GetUninstallTokensResponse
-   */
-  public async findTokensForPartialPolicyId(
+
+  public async searchTokenMetadata(
     searchString: string,
     page: number = 1,
     perPage: number = 20
-  ): Promise<GetUninstallTokensResponse> {
-    return await this.getTokensByIncludeFilter({ include: `.*${searchString}.*`, page, perPage });
+  ): Promise<GetUninstallTokensMetadataResponse> {
+    return await this.getTokenMetadataByIncludeFilter({
+      include: `.*${searchString}.*`,
+      page,
+      perPage,
+    });
   }
 
-  /**
-   * gets uninstall tokens for all policies, optionally paginated or returns all tokens
-   * @param page
-   * @param perPage
-   * @returns GetUninstallTokensResponse
-   */
-  public async getAllTokens(page?: number, perPage?: number): Promise<GetUninstallTokensResponse> {
-    return this.getTokensByIncludeFilter({ perPage, page });
+  public async getTokenMetadataForAllPolicies(
+    page?: number,
+    perPage?: number
+  ): Promise<GetUninstallTokensMetadataResponse> {
+    return this.getTokenMetadataByIncludeFilter({ perPage, page });
   }
 
-  private async getTokensByIncludeFilter({
+  private async getTokenMetadataByIncludeFilter({
     page = 1,
     perPage = SO_SEARCH_LIMIT,
     include,
   }: {
     include?: AggregationsTermsInclude;
     perPage?: number;
     page?: number;
-  }): Promise<GetUninstallTokensResponse> {
+  }): Promise<GetUninstallTokensMetadataResponse> {
+    const tokenObjects = await this.getTokenObjectsByIncludeFilter(include);
+
+    const items: UninstallTokenMetadata[] = tokenObjects
```

Comment:
I don't think I get your question. what do you mean by 'this' in your first comment?

I'm not sure about the simple search instead of using aggs: at the moment, yes, as long as we have 1 token per policy. this code, however, can handle when we have multiple tokens per policy as well as one token per policy. is this what you're thinking about?